Documentation-AM-3.20 (#948)
gravitee-io/gravitee-docs: pages/am/3.x/user-guide/mfa/user-guide-mfa-security.adoc

Now it is possible to address these circumstances by configuring and limiting the number of challenges a user can send to align with the underlying services.

The rate limit configuration is available in the `gravitee.yaml` of the AM gateway under `mfa_rate` section. The configuration is disabled by default.
The code fragment below is an enabled sample configuration which is self-explanatory. The `timeUnit` value could be `Hours`, `Minutes` or `Seconds`.

[source, yaml]
----
mfa_rate:
  enabled: true
  limit: 5
  timePeriod: 15
  timeUnit: Minutes
----

You can define these properties in the `values.yaml` file of the AM Helm Chart as well.

[source, yaml]
----
gateway:
  mfa_rate:
    enabled: true
    limit: 5
    timePeriod: 15
    timeUnit: Minutes
----

Gravitee AM monitors the MFA challenge request based on the enabled rate limit configuration and shows the user rate limit exceeded message if the user exceeds the limit.
Then the user has to wait a certain amount of time before making a successful request.

> For an example; if the limit is set to 2 for 1 minute time period and the user already has sent 2 requests, then the user must wait another 30 seconds before being able to send the 3rd request.

Author: @ashraf706 what is your view ?

Here is a screenshot of the challenge step with exhausted rate limit:

image::{% link images/am/current/graviteeio-am-userguide-mfa-ratelimit-exceed.png %}[250, 500]

NOTE: You can customise the error message by setting `mfa_challenge.rate.limit.error` property value in the *messages_en.properties* or *messages_fr.properties* file.

== Brute Force detection

> Brute force is a feature which helps to configure and limit the number of verification requests a user is allowed to send in a specific time period.

Reviewer: ```suggestion
The Brute Force Detection feature enables you to configure and limit the number of verification requests a user is allowed to send within a specific time period.
```

> This feature is configurable in domain or in application level. Follow the steps below to configure the *Brute Force* feature in domain level

Reviewer: ```suggestion
You can configure Brute Force Detection at domain level or at application level. Follow the steps below to configure the *Brute Force* feature at domain level.
```

Author: yes, exactly the same screen at application level than at domain level.

> . link:{{ '/am/current/am_userguide_authentication.html' | relative_url }}[Log in to AM Console^].

Reviewer: ```suggestion
. link:{{ '/am/current/am_userguide_authentication.html' | relative_url }}[Log in to the AM Console^].
```

. Select *User Accounts* under *SECURITY*.

> . Enable the *Brute Force Detection* in the *MFA* section.

Reviewer: ```suggestion
. Enable *Brute Force Detection* in the *MFA* section.
```

. Define your *Brute Force Detection* preferences.
+
image::{% link images/am/current/graviteeio-am-userguide-mfa-security.png %}[]

> The user will be notified with an error message once the maximum attempt is exhausted.

Reviewer: ```suggestion
Once the number of maximum attempts is reached, the user will be notified with an error message.
```

image::{% link images/am/current/graviteeio-am-userguide-mfa-brute-max.png %}[250, 500]

> A new log event `MFA_VERIFY_LIMIT_EXCEED` is also introduced to log the brute force attempt.

Reviewer: ```suggestion
A new log event, `MFA_VERIFY_LIMIT_EXCEED`, is available to log brute force attempts.
```

> NOTE: The settings can be overridden at the application level *App > Settings > accounts*.

Reviewer: ```suggestion
NOTE: You can override domain-level Brute Force Detection settings at the application level from *App > Settings > accounts*.
```

> You can customise the error message by setting `mfa_challenge.verify.limit.error` property value in the *messages_en.properties* or *messages_fr.properties* file.

Reviewer: ```suggestion
To customize the error message, modify the value of the `mfa_challenge.verify.limit.error` property in the *messages_en.properties* or *messages_fr.properties* file.
```
